doc/OVERVIEW.md

```diff
@@ -0,0 +1,469 @@
+
+ As you expect, the operations such as "enqueue" and "dequeue" have thread-safeness or cross-VM-safeness. Let's look at the *Shared Queue* related API in [hostcall.go](../proxywasm/hostcall.go):;
+
+```golang
+// DequeueSharedQueue dequeues an data from the shared queue of the given queueID.
+// In order to get queue id for a target queue, use "ResolveSharedQueue" first.
+func DequeueSharedQueue(queueID uint32) ([]byte, error)
+
+// RegisterSharedQueue registers the shared queue on this plugin context.
+// "Register" means that OnQueueReady is called for this plugin context whenever a new item is enqueued on that queueID.
+// Only available for types.PluginContext. The returned ququeID can be used for Enqueue/DequeueSharedQueue.
+// Note that "name" must be unique across all Wasm VMs which share a same "vm_id".
+// That means you can use "vm_id" can be used for separating shared queue namespace.
+//
+// Only after RegisterSharedQueue is called, ResolveSharedQueue("this vm_id", "name") succeeds
+// to retrive queueID by other VMs.
+func RegisterSharedQueue(name string) (ququeID uint32, err error)
+
+// EnqueueSharedQueue enqueues an data to the shared queue of the given queueID.
+// In order to get queue id for a target queue, use "ResolveSharedQueue" first.
+func EnqueueSharedQueue(queueID uint32, data []byte) error
+
+// ResolveSharedQueue acquires the queueID for the given vm_id and queue name.
+// The returned ququeID can be used for Enqueue/DequeueSharedQueue.
+func ResolveSharedQueue(vmID, queueName string) (ququeID uint32, err error)
+```
+
+Basically `RegisterSharedQueue` and `DequeueSharedQueue` are used by "consumer" of the queue 
+while `ResolveSharedQueue` and `EnqueueSharedQueue` are for "producer" of queue items. Note that
+
+- `RegisterSharedQueue` is used for creating a shared queue for `name` and `vm_id` of the caller. That means if you want to use a queue, then this must be called by a VM beforehand.  This can be called by `PluginContext`, and therefore we can think of "comsumers" = `PluginContext`s.
+- `ResolveSharedQueue` is used for getting the *queue id* for `name` and `vm_id`. Usually this is used by VMs that doesn't call `ResolveSharedQueue` but rather are supposed to enqueue items. This is for "producer".
+
+and both of these calls return a queue id, and it is used for `DequeueSharedQueue` and `EnqueueSharedQueue`.
+
+However, from the consumer's point of view, how can a consumer (= `PluginContext`) be notified when a queue is enqueued by a item? This is why we have the `OnQueueReady(queueID uint32)` interface in `PluginContext`. This method is called whenever an item is enqueued in a queue registered by that `PluginContext`.
+
+Also it is important to note that normally shared queues should be created by *Wasm Service* so that `OnQueueReady` or dequeue operation won't block worker threads' processing of Http or Tcp streams.
```

Comment:
does the "shared queues should be created by wasm service" only applies to the creation of the queues? Non-singleton wasm ext can use the queues, right? it's just that it's discouraged for them to **create** them?

Reply:
yeah totally. it's just discouraged to **create** them from non-singleton because otherwise OnQueueReady is called on worker threads which would block stream operations.